File: algorithms/find/binarySearch/BinarySearch.ts
class BinarySearch{
  private process(arr: number[], t: number, l: number, r: number): boolean {
    let mid: number;
    for(l; l <= r;){
        mid = (l + (r-l)/2)
      if(arr[mid] > t){
        l = mid - 1
      } else if(arr[mid] < t){
        r = mid + 1
      } else {
        return true
      }
    }
    return false
  }

  private processRecursive(arr: number[], t: number, l: number, r: number): boolean {
    let m = (l + (r - l) / 2)
    if(arr[m] > t){
      return this.processRecursive(arr, t, l, m - 1)
    } else if(arr[m] < t){
      return this.processRecursive(arr, t, m + 1, r)
    } else {
      return true
    }
  }
  
  public mainRecursive(): void {
    const arr: number[] = [22,34,66,99,124,421,576,1135,2388,4090]
    const result: boolean = this.processRecursive(arr, 66, 0, arr.length - 1)
    if(result){
      console.log("Value find")
    }else{
      throw new Error("Value not find")
    }
  }

  public main(): void {
    const arr: number[] = [22,34,66,99,124,421,576,1135,2388,4090]
    const result: boolean = this.process(arr, 66, 0, arr.length - 1)
    if(result){
      console.log("Value find")
    }else {
      throw new Error("Value not find")
    }
  }
}

(function(){
  new BinarySearch().mainRecursive()
})()

File: algorithms/find/binarySearch/BinarySearch.go
package binarysearch

import (
	"fmt"
	"log"
)

func BinarySearchIterative(arr []int, t int, l int, h int) bool {
	startIndex := l
	endindex := h

	var mid int

	for startIndex <= endindex {
		mid = int(startIndex + (endindex-startIndex)/2)
		if arr[mid] > t {
			endindex = mid - 1
		} else if arr[mid] < t {
			startIndex = mid + 1
		} else {
			return true
		}
	}
	return false
}

func Process() {
	arr := []int{22, 33, 55, 67, 77, 88, 99, 122, 323, 4556, 7657}
	result := BinarySearchIterative(arr, 4556, 0, len(arr))
	if result != true {
		log.Panic("Value not find")
	} else {
		fmt.Println("Value find")
	}
}
